src/workflow.tsx

```diff
@@ -1,12 +1,87 @@
 import * as React from 'react';
-import greet from './greet/greet';
+
+export enum Labels {
+  next = 'next',
+  previous = 'previous',
+}
+
+export interface WorkflowItem {
+  id: string;
+  next?: string;
+  previous?: string;
+  render: React.ReactNode;
+}
 
 interface WorkflowProps {
-  name: string;
+  /** the item that should be displayed when the workflow is rendered */
+  initial?: string;
+  /** the items that the workflow traverses over */
+  items: readonly WorkflowItem[];
 }
 
-const Workflow: React.FunctionComponent<WorkflowProps> = ({ name }) => (
-  <div data-testid="workflow">{greet(name)}</div>
-);
+// construct the path required to get up to, and including, the current item.
+export const buildHistory = (
+  // the current item in the workflow
+  current: WorkflowItem,
+  // items the items in the workflow
+  items: WorkflowItem[]
+): WorkflowItem[] => {
+  const go = (a: WorkflowItem, as: WorkflowItem[]): WorkflowItem[] => {
+    const p = as.find(i => i.next === a.id);
+    if (typeof p === 'undefined') {
+      return [a];
+    }
+    return [...go(p, as), { ...a, previous: p.id }];
+  };
+
+  return go(current, items);
+};
+
+const initializeWorkflow = ({
+  initial,
+  items,
+}: WorkflowProps): [WorkflowItem, WorkflowItem[]] => {
+  const [head, ...tail] = items;
+  const active = tail.find(i => i.id === initial) || head;
+  const h = buildHistory(active, [...items]);
+  const c = h.pop() as WorkflowItem;
+  return [c, h];
+};
+
+const Workflow: React.FunctionComponent<WorkflowProps> = ({
+  initial,
+  items,
+}) => {
+  const [c, h] = initializeWorkflow({ initial, items });
+  const [history, setHistory] = React.useState<WorkflowItem[]>(h);
+  const [current, setCurrent] = React.useState<WorkflowItem>(c);
```

Comment:
There is no need to specify the useState<T> type parameter because the type is inferred from c an h.

Reply:
Is there an eslint rule we could look at to enforce this?

Reply:
Not that I know, but if you are using VSCode for example, after removing the type parameters if you place to mouse over `history` you will see that the type is inferred.

Reply:
After a quick search I found https://palantir.github.io/tslint/rules/no-inferrable-types/ and https://github.com/bradzacher/eslint-plugin-typescript/blob/master/docs/rules/no-inferrable-types.md (this is archived) but I don't know how they would work really.